Add Pyro Focuser class

AnthonyHorton opened the pull request:
Adds a Pyro `Focuser` class to provide a functional `.focuser` attribute for the client side Pyro `Camera` class in order to enable direct access to remote focusers and fix #115.

Commit:
Initial implementation of Pyro Focuser class

Commit:
typo

Commit:
Avoid double import

wtgee commented on huntsman/camera/pyro.py:
```diff
@@ -92,13 +93,13 @@ def connect(self):
         (re)connect to the distributed camera.
         """
-        self.logger.debug('Connecting to {} on {}'.format(self.name, self._uri))
+        self.logger.debug('Connecting to {} at {}'.format(self.port, self._uri))
```
You're setting the `self.port` to the uri on Line 32 so here wouldn't these both be the same thing? Seems like you want to keep it at `name`.

https://github.com/AstroHuntsman/huntsman-pocs/blob/669f8831b4eba76d49e90b55e6fe3b121c93751c/huntsman/camera/pyro.py#L32

wtgee commented on huntsman/camera/pyro.py:
```diff
@@ -118,12 +119,18 @@ def connect(self):
         self._filter_type = self._proxy.filter_type
         self.logger.debug("{} connected".format(self))
 
+        if self._proxy._has_focuser:
```
Should this be the private variable or public property? I'm confusing myself to how this is set up.
```suggestion
        if self._proxy.has_focuser:
```

AnthonyHorton replied:
Public property. Of a private variable. `self._proxy` is the Pyro4 connection to the remote `CameraServer`. `CameraServer.has_focuser` is a public property because it's supposed to be accessible remotely.

wtgee commented on huntsman/focuser/pyro.py:
```diff
@@ -0,0 +1,82 @@
+
+        if not isinstance(camera, Camera):
+            msg = f"camera must be instance of huntsman.camera.pyro.Camera, got {type(camera)}."
+            raise ValueError
```
```suggestion
            raise ValueError(msg)
```

wtgee commented on huntsman/focuser/pyro.py:
```diff
@@ -0,0 +1,82 @@
+        self.model = self._proxy.focuser_model
+        self.port = self.camera.port
+        self._serial_number = self._proxy.focuser_uid
+        self.logger.debug(f"{self} connected.")
```
Does `self` come through as this Pyro object? Do we need a `__str__` method or is it fine to just see what ever is default?

AnthonyHorton replied:
`self` here is the client side Pyro `Focuser` class, which inherits from `AbstractFocuser`, which has a `__str__` method which does something sensible.

Those inherited `__str__` methods are why I was fiddling around with the `name` and `port` attributes of the Pyro `Camera` class. I decided I'd like the string representation to be closer to what it would be for an equivalent local camera.

wtgee reviewed (approved):
LGTM although I'll admit I'm a little confused as to how the camera and focus interact with each other via the proxy. Just need to spend more time looking at the whole I think.

Commit:
Avoid import error & fix attribute issues

AnthonyHorton commented:
@wtgee It is a bit convoluted. Basically I'm trying to produce the exact same interface for a distributed `Camera` _and_ its `.focuser` attribute as a local camera has, but because of the way Pyro4 works I'm doing it using a single proxy connection from the client side Pyro `Camera` class to the remote `CameraServer`. Consequently the client side Pyro `Focuser` class needs to use the proxy connection from the client side Pyro `Camera`, and the `CameraServer` needs a whole bunch more properties and methods that expose the `.focuser` of the remote `Camera`.

Commit:
Boilterplate, boilerplate everywhere

Commit:
Damn you autocomplete

Commit:
Check focuser is present for autofocus

Commit:
Ignore src

Commit:
Update test_camera to match POCS

Commit:
Update camera properties.

Commit:
Fix typo, rationalise focuser methods.

Commit:
Stop base class default values breaking stuff.

Commit:
Check is egain and bit_depth are implemented

Commit:
Too many typos

Commit:
Fix serialize bug, import error in tests

Commit:
Error importing error.

Commit:
Locally track `is_exposing` because of blocking remote exposures

Commit:
No FLIs here.

Commit:
Really no FLIs here, use internal Camera._timeout

Commit:
set exposure_event on timeout

Commit:
Hunting bugs

Commit:
Better handling of async wait errors, fix tests.

Commit:
Exception handling, better setters, fixed tests

AnthonyHorton commented:
@wtgee That escalated quickly...

I have tried to update and complete the client side Pyro `Camera` interface so that it is (almost) identical to the local camera interface in the latest version of POCS, and also added a client side `Focuser` interface so that Pyro `Camera`'s `.focuser` attribute can be used to directly control the remote focuser (almost) exactly as if it were a local focuser. I have also updated the camera tests to (almost) match those currently in POCS.

wtgee reviewed (approved):
Looks good. But ugh. Seems like a lot of work-around code and that there should be some easier solution, but I'm not sure what that would be.

AnthonyHorton commented:
@wtgee Yeah, it's not pretty. So much boilerplate code, which Pyro is supposed to minimise.

There a few reasons for the the clunkiness. One is that I'm trying to do this for a whole bunch of existing camera, focuser, etc. classes that I want to leave untouched, instead of writing new classes that are designed to be distributed. I'm also trying to reproduce the exact same interface as those existing classes on the client side (i.e. distributed objects are a drop in replacement for a local one). Pyro itself then compounds these limitations because it is designed to expose the attributes of a remote object, but it doesn't appear to have any mechanism to directly expose the attributes of the attributes of a remote object, and so on.

These factors are what pushed me to make the camera server a wrapper around a camera object instead of being a camera object itself, and to add client side wrapper classes to reproduce the original interfaces. These wrappers on both server and client side then need access methods for all of the properties and methods of the camera and focuser classes, hence all the boilerplate.

Given the freedom to made sweeping, breaking changes (and lots more time!) I could redesign things to work directly with Pyro proxies and do away with the wrapper classes entirely, eliminating almost all of the repetitive boilerplate code.
